### ISSUE TITLE
Uniquely Identifying a Yubico U2F Key

### ISSUE THREAD

**Issue by donjuan201** [Yubico/libu2f-host#45]
Please help me out with this trivial question; we have developed a whole solution around the Yubico U2F keys and used this specific lib to develop our plugins and integration components. But now we are facing a very complicated issue; we need to be able to identify a specific key (lets say when lost) and link it back to a specific user.

I know that the random challenge / response algorithm (no serial number other than the root certs serial, which is static on all our keys) does not allow for this to be done because of the randomness around it...so I am hoping that somewhere in the lib (maybe on the HID device details of any other identifier) can assist us in uniquely identifying a every key.

PLEASE PLEASE can you assist or guide us in the right direction here...

**Comment by dainnilsson** [comment]
Could you clarify the problem a bit by providing a scenario? I'm not quite following what it is you're after.

**Comment by donjuan201** [comment]
Firstly thanks for the awesomely quick response!!! Much appreciated...

Maybe I was running around a little with my previous explanation, let me try again.

Lets take a user that has enrolled on a validation server successfully. The Key Handle (maybe challenge and response as well) gets stored in the DB. Now the User looses its key and that same key is found by someone else, how does that someone else know who the key belongs to. So here is what we are trying to add to our existing solution; we would like to identify each key uniquely so that in the above case the lost key can be tracked back to the owner.

At this stage we need to loop through all the stored key handles in the DB and send them all one at a time to the authenticate method of the lib until we find the valid one. Which does work but it is a very long route to determine whose the owner of the key. Is there another unique property on the key that we can possibly use as a "serial number" like field to identify each key, maybe the device_string or device_path properties on the u2f_device object in the lib?

I hope this makes more sense, please bear with us on this...

**Comment by dainnilsson** [comment]
Thanks for clarifying! I understand what you mean now, but I have to wonder if it's a common occurrence that someone loses a key, without reporting it, and someone else finds that key and turns it in. Regardless, the only option I can think of is to loop through all key handles as you've already described, and performing a "check-only" authentication to see if the device recognizes it or not. Not the most elegant solution, but I do think it would work well enough in practice.

**Comment by conorpp** [comment]
It's worth noting that U2F tokens are by design not able to be uniquely identified for privacy reasons.

https://fidoalliance.org/specs/fido-u2f-v1.0-nfc-bt-amendment-20150514/fido-u2f-overview.html#u2f-privacy-considerations-a-recap

**Comment by candlerb** [comment]
That's exactly it: it's fundamentally impossible to identify a token with U2F.

Users can even share a token anonymously. For example, a husband and wife can use the same token, and even if they both use it to login to the same website, the website won't be able to associate the two accounts. (They each enrol the same token separately, and the website can't tell whether it's the same token or different ones)

It's a clever bit of design, but also makes U2F less useful in an enterprise environment where you *do* want to trace the tokens and where they have been used.

**Comment by larsonmpdx** [comment]
That isn't strictly true.  A website could passively query a u2f key with _all_ registrations in its database (using the non-sign challenge, which does not require a presence test).  Most (all?) u2f keys will not do any meaningful rate limiting on this type of challenge and will happily acknowledge any matched registrations.  In your example, both husband and wife registrations would come back as matching, if the website did a non-sign query for both, and that would essentially prove that the same key was used.  This is not strong cryptographic proof, because there's no signature, but it could be proof enough.  Typical client-side u2f implementations like chrome might prevent this type of querying, but that's not a guaranteed feature of u2f.

**Comment by donjuan201** [comment]
> That isn't strictly true. A website could passively query a u2f key with all registrations in its database (using the non-sign challenge, which does not require a presence test). Most (all?) u2f keys will not do any meaningful rate limiting on this type of challenge and will happily acknowledge any matched registrations. In your example, both husband and wife registrations would come back as matching, if the website did a non-sign query for both, and that would essentially prove that the same key was used. This is not strong cryptographic proof, because there's no signature, but it could be proof enough. Typical client-side u2f implementations like chrome might prevent this type of querying, but that's not a guaranteed feature of u2f.> 

This makes perfect sense; we have tried this approach on multiple occasions passing the registration key handle(s) back to the key with the "U2FH_NO_USER_PRESENCE_REQUIRED" flag command flag but the u2fh_authenticate function will not produce a response without a users presence? Maybe someone can clarify the same clarity on this for us? Below are the u2fh_cmdflags that enum that we are using with libu2f-host 1.0 library:

typedef enum
{
  U2FH_NO_USER_PRESENCE_REQUIRED = 0,
  U2FH_USER_PRESENCE_REQUIRED = 1,
  U2FH_REQUEST_NO_BLOCKING = 2
} u2fh_cmdflags;

**Comment by larsonmpdx** [comment]
querying a key with all existing registrations is more of an attack vector than an actual thing you want to do; it's not going to scale well, it's going to waste the user's bandwidth, and it's going to needlessly spread around registrations from other users (which are basically anonymized, but still)

**Comment by donjuan201** [comment]
This is only a server side process not published publicly. More for system administrators to link a key back to a user that lost a key...

**Comment by klali** [comment]
I just pushed a small fix for this where u2f-host would return U2FH_TRANSPORT_ERROR if touch was not requested. With this fix will will return U2FH_OK but no output. So the binary will exit with 0 and not print anything if a device that recognizes that keyhandle is plugged in.

**Comment by donjuan201** [comment]
EXCELLENT....we will do some testing from our side and revert back. Please confirm the released binaries that contains these updates? I see that v 1.1.1 was released yesterday?

**Comment by klali** [comment]
no, this is done after 1.1.1 and not part of any release yet (this bug was fixed today and the release of 1.1.1 was yesterday).

**Comment by candlerb** [comment]
>  A website could passively query a u2f key with all registrations in its database (using the non-sign challenge, which does not require a presence test)

Interesting - how would a website do that? Via the Low-level MessagePort API?

**Comment by smiller171** [comment]
It sounds like what you are trying to do is explicitly break the spec. If a user loses their key there should be no way to identify who's key that is, or what services it was used with. If you want this from an internal organizational perspective I would suggest physically labelling each token with a UID that you can reference in a database.

**Comment by sidsatheeesh** [comment]
Hi, I have a doubt. I have set a complete solution around the Yubico U2F keys. But now, I cannot stop duplicate registration of the same device for an user for the same app id. While checking on the keyhandles on my database they show different values for each of the duplicate registration. Please help me out.

**Comment by nevun** [comment]
We are archiving this project and closing issues. Please open a new one in the replacement library's repo: https://github.com/Yubico/libfido2 if it is still applicable.
